src/Compilers/Core/Portable/DiagnosticAnalyzer/AsyncQueue.cs

```diff
@@ -239,10 +241,30 @@ private static Task<T> WithCancellation<T>(Task<T> task, CancellationToken cance
 
             if (cancellationToken.IsCancellationRequested)
             {
-                return new Task<T>(() => default(T), cancellationToken);
+                return Task.FromCanceled<T>(cancellationToken);
             }
 
-            return task.ContinueWith(t => t, cancellationToken, TaskContinuationOptions.ExecuteSynchronously, TaskScheduler.Default).Unwrap();
+            return withCancellationSlow(task, cancellationToken);
+
+            static async Task<T> withCancellationSlow(Task<T> task, CancellationToken cancellationToken)
+            {
+                Debug.Assert(task != null);
+                Debug.Assert(cancellationToken.CanBeCanceled);
```

Comment:
`Debug.Assert` not needed

Reply:
These assertions were a direct translation of https://github.com/Microsoft/vs-threading/blob/a8e7917134d37bd8bda4f3b476d25c55ffddad9e/src/Microsoft.VisualStudio.Threading/ThreadingTools.cs#L308-L309. I feel they help clarify preconditions that the method relies on for correctness. Are they problematic?

Reply:
It's not that they're problematic, they're just redundant. The outer method already has the exact same assertions.

---
In reply to: [267996984](https://github.com/dotnet/roslyn/pull/34342#discussion_r267996984) [](ancestors = 267996984)